data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java

```diff
@@ -538,21 +540,109 @@ public JacksonEvent build() {
+            } else if (getExcludeKeys() != null && !getExcludeKeys().isEmpty()) {
+                jsonString = searchAndFilter(getBaseNode(), "", getExcludeKeys(), EXCLUDE_ALL);
+            } else if (getBaseNode() !=event.getJsonNode()) {
+                jsonString = event.getAsJsonString(getRootKey());
+            } else {
+                // Some successors have its own implementation of toJsonString, such as JacksonSpan.
+                // In such case, it's only used when the root key is not provided.
+                // TODO: Need to check if such behaviour is expected.
+                jsonString = event.toJsonString();
+            }
+
             final String tagsKey = getTagsKey();
-            if(tagsKey != null) {
+            if (tagsKey != null) {
                 final JsonNode tagsNode = mapper.valueToTree(event.getMetadata().getTags());
-                return jsonString.substring(0, jsonString.length()-1) + ",\""+tagsKey+"\":" + tagsNode.toString()+"}";
+                return jsonString.substring(0, jsonString.length() - 1) + ",\"" + tagsKey + "\":" + tagsNode.toString() + "}";
             }
             return jsonString;
         }
+
+        /**
+         * Perform DFS(Depth-first search) like traversing using recursion on the Json Tree and return the json string.
+         * This supports filtering (to include or exclude) from a list of keys.
+         *
+         * @param node         Root node to start traversing
+         * @param path         Json path, e.g. /foo/bar
+         * @param filterKeys   A list of filtered keys
+         * @param filterAction Either to include (RETAIN_ALL or true) or to exclude (EXCLUDE_ALL or false)
+         * @return a json string with filtered keys
+         */
+        String searchAndFilter(JsonNode node, String path, final List<String> filterKeys, boolean filterAction) {
+
+            if (node.isArray()) { // for array node.
+                StringJoiner sj = new StringJoiner(",", "[", "]");
+                node.forEach(childNode -> sj.add(searchAndFilter(childNode, path, filterKeys, filterAction)));
+                return sj.toString();
+            } else {
+                StringJoiner sj = new StringJoiner(",", "{", "}");
+                List<String> valueList = new ArrayList<>();
+
+                node.properties().forEach(entry -> {
+                    String keyPath = path + SEPARATOR + entry.getKey();
+                    // Track whether the key is found in the filter list.
+                    // Different behaviours between include and exclude action.
+                    boolean found = false;
+                    for (String key : filterKeys) {
+                        if (keyPath.equals(key)) {
+                            found = true;
+                            // To keep the order.
+                            if (filterAction == RETAIN_ALL) {
+                                valueList.add("\"" + entry.getKey() + "\":" + entry.getValue().toString());
```

Comment:
Are we sure that `entry.gevValue().toString()` provides the right value here? What if the string has double quotes in it? Are they being escaped? Any other special characters?

Reply:
Yes, `entry.getValue().toString()` is the right value here.  All the special characters are already taken care of by this.  The output could be a number (e.g. 10), a character with double quotes (e.g. "ye\"llow"),  or even the full json tree (e.g. ["msg 1","msg 2","msg 3"], [{"name":"abc","age":"bcd"},{"name":"hello","age":"world"}] etc.